#PBS -N mmseqs
#PBS -l nodes=1:ppn=2
#PBS -l mem=10gb
#PBS -l walltime=12:00:00
#PBS -q inferno     
#PBS -A GT-ktk3-CODA20
#PBS -o 00a_log/02c_mmseqs_${n}.out
#PBS -e 00a_log/02c_mmseqs_${n}.err

# Change to working directory
cd $PBS_O_WORKDIR

# make a tmp folder
mkdir ${odir}/tmp

# create an mmseqs database
DB=02_mmseqsDB
mkdir ${odir}/$DB
mmseqs createdb $infile ${odir}/$DB/$DB --dbtype 1

# Cluster at 90% nucleotide ID
C90=02_mmseqs_results
mkdir ${odir}/$C90
mmseqs cluster ${odir}/$DB/$DB ${odir}/$C90/$C90 tmp --min-seq-id 0.90 --cov-mode 1 -c 0.5 --cluster-mode 2 --cluster-reassign --threads 2
# convert to tsv format
mmseqs createtsv ${odir}/$DB/$DB ${odir}/$DB/$DB ${odir}/$C90/$C90 ${odir}/$C90.tsv --threads 2
# for amino acid clustering get the representative sequence for annotation.
C90rep=02_mmseqs_reps
mkdir ${odir}/$C90rep
mmseqs createsubdb ${odir}/$C90/$C90 ${odir}/$DB/$DB ${odir}/$C90rep/$C90rep
mmseqs convert2fasta ${odir}/$C90rep/$C90rep ${odir}/${C90rep}.fasta

rm -r ${odir}/tmp

## Run Log ##
## qsub -v infile=,odir=,n= ../00b_PBS/07c_MMSeqs2_Cluster.pbs
